File: python-algorithm/algo_09_math/L66.cpp
//
// Created by 李恒 on 2020/6/30.
//

#include <iostream>
#include <vector>
#include <stack>
using namespace std;

class Solution {
public:
    vector<int> plusOne(vector<int>& digits) {
        stack<int> stk;
        int sz = digits.size();
        int flag = 1;
        for(int i=sz-1; i>=0; --i){
            int m = (digits[i]+flag)%10;
            flag = (digits[i]+flag)/10;
            stk.push(m);
        }
        if(flag == 1)
            stk.push(1);
        vector<int> res;
        while(!stk.empty()){
            res.push_back(stk.top());
            stk.pop();
        }
        return res;
    }
};

File: python-algorithm/algo_09_math/L66.py
# -*- coding: utf-8 -*-
# @Time    : 2020/10/6 13:36
# @Author  : No Name

class Solution:
    def plusOne(self, digits: List[int]) -> List[int]:
        res = [digits[i] for i in range(len(digits))]
        # res = digits
        carry = 1
        for i in range(len(digits)-1, -1, -1):
            tmp = digits[i] + carry
            res[i] = tmp%10
            carry = tmp//10
            # 剪枝，提前终止
            if carry == 0:
                return res
        if carry == 1:
            res.insert(0,1)
        return res